### PR TITLE
LibWeb: Update HTMLScriptElement prepare and execute to latest HTML spec

### PR DESCRIPTION
**Pull request by awesomekling**
The net result is some nice simplification of the prepare algorithm and a bit of initial scaffolding for import maps support.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp [hunk 1]
```diff
@@ -49,236 +50,264 @@ void HTMLScriptElement::begin_delaying_document_load_event(DOM::Document& docume
...
     String script_block_type;
-    bool has_type = has_attribute(HTML::AttributeNames::type);
-    bool has_language = has_attribute(HTML::AttributeNames::language);
-    if ((has_type && attribute(HTML::AttributeNames::type).is_empty())
-        || (!has_type && has_language && attribute(HTML::AttributeNames::language).is_empty())
-        || (!has_type && !has_language)) {
+    bool has_type_attribute = has_attribute(HTML::AttributeNames::type);
+    bool has_language_attribute = has_attribute(HTML::AttributeNames::language);
+    if ((has_type_attribute && attribute(HTML::AttributeNames::type).is_empty())
+        // then let the script block's type string for this script element be "text/javascript".
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp:163]: would it make more sense for this comment be above the assignment to script block type?

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp [hunk 2]
```diff
@@ -49,236 +50,264 @@ void HTMLScriptElement::begin_delaying_document_load_event(DOM::Document& docume
...
-    } else if (has_type) {
-        // Otherwise, if the script element has a type attribute, let the script block's type string for this script element be the value of that attribute.
-        script_block_type = attribute(HTML::AttributeNames::type);
-    } else if (!attribute(HTML::AttributeNames::language).is_empty()) {
-        // Otherwise, the element has a non-empty language attribute; let the script block's type string for this script element be the concatenation of the string "text/" followed by the value of the language attribute.
+    }
+    // Otherwise, if el has a type attribute,
+    else if (has_type_attribute) {
+        // then let the script block's type string be the value of that attribute with leading and trailing ASCII whitespace stripped.
+        script_block_type = attribute(HTML::AttributeNames::type).trim_whitespace();
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp:171]: Linus added some fancy helpers to LibWeb/Infra for things like `https://infra.spec.whatwg.org/#strip-leading-and-trailing-ascii-whitespace` which is used below on line 180 `script_block_type.trim(Infra::ASCII_WHITESPACE)` which should probably be used here for consistency

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp [hunk 3]
```diff
@@ -49,236 +50,264 @@ void HTMLScriptElement::begin_delaying_document_load_event(DOM::Document& docume
...
-        // - If neither of the above conditions are true, then return. No script is executed.
+    }
+    // 11. Otherwise, if the script block's type string is an ASCII case-insensitive match for the string "importmap",
+    else if (script_block_type.equals_ignoring_case("importmap"sv)) {
+        // then set el's type to "importmap".
+        m_script_type = ScriptType::ImportMap;
+    }
+    // 12. Otherwise, return. (No script is executed, and el's type is left as null.)
+    else {
         dbgln("HTMLScriptElement: Refusing to run script because the type '{}' is not recognized.", script_block_type);
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp:196]: worth a `VERIFY(m_script_type == ScriptType::Null)` here? Although saying that, we don't have an "Invalid" state for that enum :), which is against da shpec

> A [script](https://html.spec.whatwg.org/multipage/scripting.html#the-script-element) element has a type, which is either null, "classic", "module", or "importmap", initially null. It is determined when the element is [prepared](https://html.spec.whatwg.org/multipage/scripting.html#prepare-the-script-element), based on the [type](https://html.spec.whatwg.org/multipage/scripting.html#attr-script-type) attribute of the element at that time.

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp [hunk 4]
```diff
@@ -287,148 +316,183 @@ void HTMLScriptElement::prepare_script()
...
+        auto base_url = document().base_url();
 
-        // 2. Switch on the script's type:
+        // 2. Switch on el's type:
+        // -> "classic"
         if (m_script_type == ScriptType::Classic) {
-            // -> "classic"
             // 1. Let script be the result of creating a classic script using source text, settings object, base URL, and options.
             // FIXME: Pass options.
             auto script = ClassicScript::create(m_document->url().to_string(), source_text, document().relevant_settings_object(), base_url, m_source_line_number);
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp:363]: pass `settings_object` here, like on 343?

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp [hunk 5]
```diff
@@ -393,10 +393,10 @@ void HTMLScriptElement::prepare_script()
     }
 
     // 32. If el's type is "classic" and el has a src attribute, or el's type is "module":
-    if ((m_script_type == ScriptType::Classic && has_attribute(HTML::AttributeNames::src))
-        || m_script_type == ScriptType::Module) {
+    if ((m_script_type == ScriptType::Classic && has_attribute(HTML::AttributeNames::src)) || m_script_type == ScriptType::Module) {
         // 1. Assert: el's result is "uninitialized".
-        VERIFY(m_result.has<ResultState::Uninitialized>());
+        // FIXME: I believe this step to be a spec bug, as el's result could be "null" after step 31 for "module" above.
+        //        Waiting to hear back from WHATWG.
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp:399]: Is there a whatwg github Issue number we can add here?

**Review by ADKaster — COMMENTED** [review]
Neat! I found some nitpicks

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.h [hunk 6]
```diff
@@ -64,28 +64,44 @@ class HTMLScriptElement final
...
     Function<void()> m_script_ready_callback;
 
+    // https://html.spec.whatwg.org/multipage/scripting.html#concept-script-result
     JS::GCPtr<Script> m_script;
 
+    // https://html.spec.whatwg.org/multipage/scripting.html#concept-script-delay-load
     Optional<DOM::DocumentLoadEventDelayer> m_document_load_event_delayer;
 
+    bool m_script_ready { false };
+    bool m_failed_to_load { false };
```
[the remark this answers is not in the document]
Comment (awesomekling) [reply Userland/Libraries/LibWeb/HTML/HTMLScriptElement.h:114]: Sure, let's not move them.

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp (hunk 1 above)
Comment (awesomekling) [reply Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp:163]: ....yes 😅

File: Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp (hunk 3 above)
Comment (awesomekling) [reply Userland/Libraries/LibWeb/HTML/HTMLScriptElement.cpp:196]: Great point! Will add `ScriptType::Null`

**Review by ADKaster — APPROVED** [review]

**Pull request merged**